In [1]:
import pandas as pd
import os

# Ignorar todos los warnings
import warnings
warnings.filterwarnings("ignore")

In [2]:
path="C:/Users/galla/Dropbox/PC/Documents/1 Pedro G Gallardo/Master IA y Big Data/Casa/proyecto tesla/ProyectoTesla/data/data_prepared"
os.chdir(path)

In [3]:
df_combined = pd.read_csv('data_prepared.csv', sep=';')
df_combined

kilometraje  precio   año  modelo_Model 3  modelo_Model S  \
0            2622   82010  2024               0               1   
1            6196   88990  2024               0               1   
2             792   90610  2024               0               1   
3             551   90770  2024               0               1   
4             881   91500  2024               0               1   
...           ...     ...   ...             ...             ...   
1098            0   59980  2024               1               0   
1099           28  112790  2024               0               0   
1100           18  115490  2024               0               1   
1101           26  116120  2024               0               1   
1102           20  117590  2024               0               0   

      modelo_Model Y  color_azul  color_blanco  color_gris  color_negro  \
0                  0           0             1           0            0   
1                  0           1             0           0            0   
2                  0           1             0           0            0   
3                  0           0             0           0            1   
4                  0           0             0           1            0   
...              ...         ...           ...         ...          ...   
1098               0           0             0           1            0   
1099               0           0             0           1            0   
1100               0           0             0           1            0   
1101               0           0             0           1            0   
1102               0           0             0           1            0   

      color_plateado  color_rojo  pais_España  color_granate  color_marron  \
0                  0           0            1              0             0   
1                  0           0            1              0             0   
2                  0           0            1              0             0   
3                  0           0            1              0             0   
4                  0           0            1              0             0   
...              ...         ...          ...            ...           ...   
1098               0           0            0              0             0   
1099               0           0            0              0             0   
1100               0           0            0              0             0   
1101               0           0            0              0             0   
1102               0           0            0              0             0   

      color_plata  modelo_Model X  pais_Francia  pais_Portugal  
0               0               0             0              0  
1               0               0             0              0  
2               0               0             0              0  
3               0               0             0              0  
4               0               0             0              0  
...           ...             ...           ...            ...  
1098            0               0             1              0  
1099            0               1             1              0  
1100            0               0             1              0  
1101            0               0             1              0  
1102            0               1             1              0  

[1103 rows x 19 columns]

In [4]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

In [7]:
cols = list(df_combined.columns)
if "precio" in cols:
    cols.remove("precio")
cols

['kilometraje',
 'año',
 'modelo_Model 3',
 'modelo_Model S',
 'modelo_Model Y',
 'color_azul',
 'color_blanco',
 'color_gris',
 'color_negro',
 'color_plateado',
 'color_rojo',
 'pais_España',
 'color_granate',
 'color_marron',
 'color_plata',
 'modelo_Model X',
 'pais_Francia',
 'pais_Portugal']

In [8]:
X = df_combined[cols]
y =df_combined['precio']

In [9]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [10]:
modelo = RandomForestRegressor(random_state=2517)

In [11]:
modelo.fit(X_train, y_train)

RandomForestRegressor(random_state=2517)

In [12]:
y_predict = modelo.predict(X_test)

In [13]:
mse = mean_squared_error(y_test, y_predict)
print(f"MSE: {mse}")

MSE: 149376996.41810498


In [14]:
import numpy as np
rmse = np.sqrt(mse)
print(f"RMSE: {rmse}")

RMSE: 12221.988235066543


In [15]:
mae = mean_absolute_error(y_test, y_predict)
print(f"MAE: {mae}")

MAE: 6603.321530350054


El valor 𝑅2 varía entre −∞ y 1:
- R2 = 1: predicción perfecta.
- R2 = 0: el modelo es tan bueno como usar la media.
- R2 < 0: el modelo es peor que usar la media.

In [16]:
r2 = r2_score(y_test, y_predict)
print(f"R2 Score: {r2}")

R2 Score: 0.7129127387718319


In [17]:
from tqdm import tqdm

# Supongamos que ya tienes definidos X_train, y_train, X_test, y_test
n_arboles = [1, 2, 4, 5, 10, 20, 30, 40, 50, 100, 150, 200, 250, 300, 350, 400, 450, 500, 600, 700, 800, 900, 1000]
criterios = ['squared_error', 'absolute_error', 'friedman_mse', 'poisson']
mejor_error = 99999999

# Para guardar el MSE promedio para cada n_estimators
mse_promedios = []

for i in tqdm(n_arboles):
    mse_valores = []  # Guardar los MSE para diferentes max_depth con el mismo n_estimators
    for crit in criterios:   # Cambiar criterios
        for j in range(1, 20):  # Cambiando max_depth
            regr = RandomForestRegressor(n_estimators=i, criterion=crit, max_depth=j, random_state=2517)
            regr.fit(X_train, y_train)
            y_predicho = regr.predict(X_test)
            mse = mean_squared_error(y_test, y_predicho)

            mse_valores.append(mse)  # Guardar el MSE actual

            if mse < mejor_error:
                mejor_error = mse
                print(f'Random Forest - MSE: {mse}\nn_estimators: {i}\ncriterion: {crit}\nmax_depth: {j}\nrandom_state: 2517\n\n')
    
    # Calcular el promedio del MSE para este valor de n_estimators
    mse_promedio = sum(mse_valores) / len(mse_valores)
    mse_promedios.append(mse_promedio)


  0%|          | 0/23 [00:00<?, ?it/s]

Random Forest - MSE: 86470778.65010877
n_estimators: 1
criterion: squared_error
max_depth: 3
random_state: 2517


Random Forest - MSE: 78992519.42760219
n_estimators: 1
criterion: squared_error
max_depth: 5
random_state: 2517




  9%|▊         | 2/23 [00:01<00:19,  1.09it/s]

Random Forest - MSE: 78008268.55786136
n_estimators: 4
criterion: squared_error
max_depth: 4
random_state: 2517


Random Forest - MSE: 74962585.44956994
n_estimators: 4
criterion: squared_error
max_depth: 5
random_state: 2517


Random Forest - MSE: 72925986.77439196
n_estimators: 4
criterion: absolute_error
max_depth: 9
random_state: 2517




 13%|█▎        | 3/23 [00:03<00:27,  1.36s/it]

Random Forest - MSE: 71361225.99158372
n_estimators: 5
criterion: absolute_error
max_depth: 9
random_state: 2517




100%|██████████| 23/23 [55:16<00:00, 144.21s/it]


Mejor metrica:
Random Forest - MSE: 71361225.99158372
- n_estimators: 5
- criterion: absolute_error
- max_depth: 9
- random_state: 2517

In [18]:
rmse = np.sqrt(mse)
print(f"RMSE: {rmse}")

RMSE: 11853.030291641178


In [19]:
y_predicho = regr.predict(X_test)
r2_valor = r2_score(y_test, y_predict)
print(f"R2 Score: {r2_valor}")

R2 Score: 0.7129127387718319


In [20]:
# Supongamos que ya tienes definidos X_train, y_train, X_test, y_test
n_arboles = [1, 2, 4, 5, 10, 20, 30, 40, 50, 100, 150, 200, 250, 300, 350, 400, 450, 500]
criterios = ['squared_error', 'absolute_error', 'friedman_mse', 'poisson']
mejor_error = 99999999

# Para guardar el MSE promedio para cada n_estimators
rmse_promedios = []

for i in tqdm(range(1, 20)):
    rmse_valores = []  # Guardar los MSE para diferentes max_depth con el mismo n_estimators
    for crit in criterios:   # Cambiar criterios
        for j in range(1, 15):  # Cambiando max_depth
            regr2 = RandomForestRegressor(n_estimators=i, criterion=crit, max_depth=j, random_state=2517)
            regr2.fit(X_train, y_train)
            y_predicho = regr2.predict(X_test)
            mse = mean_squared_error(y_test, y_predicho)
            rmse = np.sqrt(mse)

            rmse_valores.append(rmse)  # Guardar el MSE actual

            if rmse < mejor_error:
                mejor_error = rmse
                print(f'Random Forest - RMSE: {rmse}\nn_estimators: {i}\ncriterion: {crit}\nmax_depth: {j}\nrandom_state: 2517\n\n')
    
    # Calcular el promedio del MSE para este valor de n_estimators
    rmse_promedio = sum(rmse_valores) / len(rmse_valores)
    rmse_promedios.append(rmse_promedio)


  0%|          | 0/19 [00:00<?, ?it/s]

Random Forest - RMSE: 17467.508209500567
n_estimators: 1
criterion: squared_error
max_depth: 1
random_state: 2517


Random Forest - RMSE: 13921.016632348159
n_estimators: 1
criterion: squared_error
max_depth: 2
random_state: 2517


Random Forest - RMSE: 9298.966536670017
n_estimators: 1
criterion: squared_error
max_depth: 3
random_state: 2517


Random Forest - RMSE: 8887.77359227845
n_estimators: 1
criterion: squared_error
max_depth: 5
random_state: 2517




 11%|█         | 2/19 [00:01<00:09,  1.71it/s]

Random Forest - RMSE: 8870.515419545916
n_estimators: 3
criterion: squared_error
max_depth: 5
random_state: 2517




 16%|█▌        | 3/19 [00:02<00:12,  1.32it/s]

Random Forest - RMSE: 8832.228968831218
n_estimators: 4
criterion: squared_error
max_depth: 4
random_state: 2517


Random Forest - RMSE: 8658.093638299943
n_estimators: 4
criterion: squared_error
max_depth: 5
random_state: 2517


Random Forest - RMSE: 8539.671350490718
n_estimators: 4
criterion: absolute_error
max_depth: 9
random_state: 2517




 21%|██        | 4/19 [00:03<00:14,  1.06it/s]

Random Forest - RMSE: 8447.557397945498
n_estimators: 5
criterion: absolute_error
max_depth: 9
random_state: 2517




100%|██████████| 19/19 [00:53<00:00,  2.80s/it]


In [21]:
y_predicho = regr2.predict(X_test)
r2_valor = r2_score(y_test, y_predict)
print(f"R2 Score: {r2_valor}")

R2 Score: 0.7129127387718319


Mejor metrica RMSE (antiguo)
- Random Forest - RMSE: 8643.158966517014
- n_estimators: 2
- criterion: poisson
- max_depth: 6
- random_state: 2517